ambry-tools/src/main/java/com.github.ambry/store/StoreCopier.java

```diff
@@ -0,0 +1,278 @@
+        storeCopier.copy(new StoreFindTokenFactory(storeKeyFactory).getNewFindToken());
+      }
+    }
+  }
+
+  /**
+   * @param storeId the name/id of the {@link Store}.
+   * @param srcDir the directory of the {@link Store} to be copied from
+   * @param tgtDir the directory of the {@link Store} to be copied to.
+   * @param storeCapacity the capacity of the store.
+   * @param fetchSizeInBytes the size of each fetch from the soure store.
+   * @param storeConfig {@link StoreConfig} that contains config to initiate a {@link BlobStore}.
+   * @param metricRegistry {@link MetricRegistry} to use for metrics.
+   * @param storeKeyFactory the {@link StoreKeyFactory} to use for {@link StoreKey}s in the {@link Store}.
+   * @param diskIOScheduler the {@link DiskIOScheduler} to use.
+   * @param transformers the list of {@link Transformer} functions to execute. They will be executed in order.
+   * @param time the {@link Time} instance to use.
+   * @throws StoreException
+   */
+  public StoreCopier(String storeId, File srcDir, File tgtDir, long storeCapacity, long fetchSizeInBytes,
+      StoreConfig storeConfig, MetricRegistry metricRegistry, StoreKeyFactory storeKeyFactory,
+      DiskIOScheduler diskIOScheduler, List<Transformer> transformers, Time time) throws StoreException {
+    this.storeId = storeId;
+    this.fetchSizeInBytes = fetchSizeInBytes;
+    this.transformers = transformers;
+    StorageManagerMetrics metrics = new StorageManagerMetrics(metricRegistry);
+    MessageStoreRecovery recovery = new BlobStoreRecovery();
+    src = new BlobStore(storeId, storeConfig, null, null, diskIOScheduler, metrics, srcDir.getAbsolutePath(),
+        storeCapacity, storeKeyFactory, recovery, null, time);
+    tgt = new BlobStore(storeId + "_tmp", storeConfig, scheduler, null, diskIOScheduler, metrics,
+        tgtDir.getAbsolutePath(), storeCapacity, storeKeyFactory, recovery, null, time);
+    src.start();
+    tgt.start();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (!isOpen) {
+      return;
+    }
+    try {
+      scheduler.shutdown();
+      if (!scheduler.awaitTermination(1, TimeUnit.SECONDS)) {
+        logger.error("Could not shut down scheduler");
+      }
+      src.shutdown();
+      tgt.shutdown();
+      isOpen = false;
+    } catch (Exception e) {
+      throw new IOException(e);
+    }
+  }
+
+  /**
+   * Copies data starting from {@code startToken} until all the data is copied.
+   * @param startToken the {@link FindToken} to start copying from. Does not perform any duplication checks at
+   *                   destination.
+   * @return the {@link FindToken} until which data has been copied.
+   * @throws IOException if there is any I/O error while copying.
+   * @throws StoreException if there is any exception dealing with the stores.
+   */
+  public FindToken copy(FindToken startToken) throws IOException, StoreException {
```

Comment:
How can we ensure that this tool evolves along with changes to the store code and logic, as some of the logic is being borrowed here (and not reused). This applies to other store tools as well, but it is more important here as this tool is used to change state of running systems. It was okay before to realize that, say, DumpIndexTool was broken - we would investigate and fix it. It is not okay for this tool to be broken silently when we run it. Unlike other tools, this one is a first class citizen.

Reply:
Will chat with you about this - want to understand what the borrowed code is since I feel like there isn't any store code here.